api/layer/multipart_upload.go

```diff
@@ -589,9 +596,90 @@ func (n *layer) reUploadFollowingParts(ctx context.Context, uploadParams UploadP
 	for _, part := range parts {
 		uploadParams.PartNumber = part.Number
 
-		if err = n.reUploadPart(ctx, uploadParams, part.OID, bktInfo, multipartInfo); err != nil {
-			return fmt.Errorf("reupload number=%d: %w", part.Number, err)
+		if len(part.Elements) > 0 {
+			if err = n.reUploadSegmentedPart(ctx, uploadParams, part, bktInfo, multipartInfo); err != nil {
+				return fmt.Errorf("reupload number=%d: %w", part.Number, err)
+			}
+		} else {
+			if err = n.reUploadPart(ctx, uploadParams, part.OID, bktInfo, multipartInfo); err != nil {
+				return fmt.Errorf("reupload number=%d: %w", part.Number, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (n *layer) reUploadSegmentedPart(ctx context.Context, uploadParams UploadPartParams, part *data.PartInfo, bktInfo *data.BucketInfo, multipartInfo *data.MultipartInfo) error {
+	var (
+		eg                     errgroup.Group
+		pipeReader, pipeWriter = io.Pipe()
+	)
+
+	eg.Go(func() error {
+		var (
+			err        error
+			elementObj *object.Object
+		)
+
+		for _, element := range part.Elements {
+			elementObj, err = n.objectGet(ctx, bktInfo, element.OID)
+			if err != nil {
+				err = fmt.Errorf("get part oid=%s, element oid=%s: %w", part.OID.String(), element.OID.String(), err)
+				break
+			}
+
+			if _, err = pipeWriter.Write(elementObj.Payload()); err != nil {
+				err = fmt.Errorf("write part oid=%s, element oid=%s: %w", part.OID.String(), element.OID.String(), err)
+				break
+			}
+
+			// The part contains all elements for Split chain and contains itself as well.
+			// We mustn't remove it here, it will be removed on MultipartComplete.
+			if part.OID == element.OID {
+				continue
+			}
+
+			if deleteErr := n.objectDelete(ctx, bktInfo, element.OID); deleteErr != nil {
+				n.log.Error(
+					"couldn't delete object",
+					zap.Error(deleteErr),
+					zap.String("cnrID", bktInfo.CID.EncodeToString()),
+					zap.String("uploadID", multipartInfo.UploadID),
+					zap.Int("partNumber", part.Number),
+					zap.String("part.OID", part.OID.String()),
+					zap.String("part element OID", element.OID.String()),
+				)
+				// no return intentionally.
+			}
+		}
+
+		if pipeCloseErr := pipeWriter.Close(); pipeCloseErr != nil {
```

Comment:
So if `pipeCloseErr == nil` and `err != nil` we're fine? I want to deduplicate `Close()`, but errors should be passed properly still.

Reply:
Updated